### PR TITLE
Issue/1577 image upload placeholder previews

### LINKED ISSUE #1577: Show previews in uploading product images
Right now we simply show progress indicators for uploading images. We should instead show a preview of the uploading image, with a alpha applied to make it appear dimmer.

### PR DESCRIPTION
**Pull request by nbradbury**
Closes #1577 - adds a preview of uploading images, and updates the upload notification to show the number of images being uploaded and the upload progress for each image.

One change I made in this PR was that previously multiple uploads were dispatched concurrently, but now they're uploaded in order. This enables showing an accurate upload process, and more importantly the previous approach seemed too resource-intense.

placeholder

notif

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCProductImageGalleryView.kt [hunk 1]
```diff
@@ -105,10 +106,19 @@ class WCProductImageGalleryView @JvmOverloads constructor(
...
-    fun setPlaceholderCount(count: Int) {
-        adapter.setPlaceholderCount(count)
+    fun setPlaceholderImageUris(imageUriList: List<Uri>) {
+        val placeholders = ArrayList<WCProductImageModel>()
+        for (index in imageUriList.indices) {
+            // use a negative id so we can check it in isPlaceholder() below
+            val id = (-index - 1).toLong()
+            placeholders.add(0, WCProductImageModel(id).also {
+                // set the image src to this uri so we can preview it while uploading
+                it.src = imageUriList[index].toString()
```
Comment (anitaa1990) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCProductImageGalleryView.kt]: Very nice touch <3

**Comment by nbradbury** [comment]
> Would it make sense to wait till all the images are uploaded (either success/error) before making the API call to update the product?

I spent some time on this, and the problem is that we need to assign the uploaded media to the product right after each upload so the user can see each successfully uploaded image immediately after the upload. If we wait to assign the media until after the last upload completes, the user wouldn't see the uploaded media until the end.

**Comment by anitaa1990** [comment]
> I spent some time on this, and the problem is that we need to assign the uploaded media to the product right after each upload so the user can see each successfully uploaded image immediately after the upload. If we wait to assign the media until after the last upload completes, the user wouldn't see the uploaded media until the end.

Ah, that's a good point! Do you think we could have a cached product model in the `ViewModel` and update the `mediaId` to that instead of updating it to the API one at a time? It's not a big deal but from what I remember, it looks like we would need to cache all the product changes in memory and only update the backend when user clicks on `Publish` in the product detail screen. I'm working on this for product description but still a work in progress (trying to come up with a solution that could work for all product fields and not just product description). Do you think this approach could work for Product images too or would that be making it too complicated?

**Comment by nbradbury** [comment]
> Do you think we could have a cached product model in the `ViewModel` and update the `mediaId` to that instead of updating it to the API one at a time? It's not a big deal but from what I remember, it looks like we would need to cache all the product changes in memory and only update the backend when user clicks on `Publish` in the product detail screen.

You're right - at some point we will have to cache the changes to product images. But I don't think we want to do that when adding images to existing products. Consider the case that the user starts uploading several images. If we rely on the ViewModel to cache the image changes, what happens  if they leave product detail while the upload is happening and then return to the detail?

**Pull request merged**